File: sdk/python/velocity_sdk/payload_codec.py
"""
VELOCITY-WorkFlow Python SDK - Payload encoding/decoding.

Provides codecs for serializing workflow payloads (JSON, protobuf, raw bytes).
"""

import json
from typing import Any, Protocol
from dataclasses import dataclass


class PayloadCodec(Protocol):
    """Interface for payload encoding/decoding."""

    def encode(self, data: Any) -> bytes:
        """Encode data to bytes."""
        ...

    def decode(self, data: bytes) -> Any:
        """Decode bytes to data."""
        ...


@dataclass
class JsonCodec:
    """JSON payload codec."""

    ensure_ascii: bool = False
    sort_keys: bool = False

    def encode(self, data: Any) -> bytes:
        """Encode data as JSON bytes."""
        return json.dumps(
            data,
            ensure_ascii=self.ensure_ascii,
            sort_keys=self.sort_keys,
        ).encode("utf-8")

    def decode(self, data: bytes) -> Any:
        """Decode JSON bytes to Python object."""
        return json.loads(data.decode("utf-8"))


@dataclass
class BinaryCodec:
    """Raw binary codec (passthrough)."""

    def encode(self, data: bytes) -> bytes:
        """Return data as-is."""
        if not isinstance(data, (bytes, bytearray)):
            raise TypeError(f"BinaryCodec expects bytes, got {type(data).__name__}")
        return bytes(data)

    def decode(self, data: bytes) -> bytes:
        """Return data as-is."""
        return data


class ProtobufCodec:
    """Protocol Buffers codec (requires protobuf library)."""

    def __init__(self, message_type: Any = None):
        """
        Initialize with a protobuf message type.

        Args:
            message_type: Protobuf message class (e.g., MyMessage)
        """
        self.message_type = message_type

    def encode(self, data: Any) -> bytes:
        """Encode protobuf message to bytes."""
        if self.message_type is None:
            raise ValueError("message_type must be set for ProtobufCodec")

        if hasattr(data, "SerializeToString"):
            return data.SerializeToString()

        # Try to create message from dict
        msg = self.message_type()
        if isinstance(data, dict):
            for key, value in data.items():
                setattr(msg, key, value)
        else:
            msg.CopyFrom(data)

        return msg.SerializeToString()

    def decode(self, data: bytes) -> Any:
        """Decode bytes to protobuf message."""
        if self.message_type is None:
            raise ValueError("message_type must be set for ProtobufCodec")

        msg = self.message_type()
        msg.ParseFromString(data)
        return msg


class CodecChain:
    """Chain multiple codecs together (e.g., JSON -> compression)."""

    def __init__(self, codecs: list[PayloadCodec]):
        if not codecs:
            raise ValueError("CodecChain requires at least one codec")
        self.codecs = codecs

    def encode(self, data: Any) -> bytes:
        """Encode through all codecs in order."""
        result = data
        for codec in self.codecs:
            result = codec.encode(result)
        return result if isinstance(result, bytes) else self.codecs[-1].encode(result)

    def decode(self, data: bytes) -> Any:
        """Decode through all codecs in reverse order."""
        result: Any = data
        for codec in reversed(self.codecs):
            result = codec.decode(result)
        return result

File: sdk/ruby/lib/velocity_sdk/payload_codec.rb
# frozen_string_literal: true

require 'json'

module VelocitySdk
  # Interface for payload encoding and decoding.
  #
  # Implementations must handle serialization of arbitrary data to bytes
  # and deserialization from bytes back to the original data format.
  module PayloadCodec
    # Encode data to bytes.
    # @param data [Object] The data to encode
    # @return [String] Encoded bytes (binary string)
    # @raise [CodecError] if encoding fails
    def encode(data)
      raise NotImplementedError
    end

    # Decode bytes to data.
    # @param data [String] Raw bytes to decode
    # @return [Object] Decoded data
    # @raise [CodecError] if decoding fails
    def decode(data)
      raise NotImplementedError
    end
  end

  # Error raised when codec operations fail.
  class CodecError < StandardError; end

  # JSON payload codec.
  #
  # Uses Ruby's built-in JSON library for serialization.
  class JsonCodec
    include PayloadCodec

    def encode(data)
      return '' if data.nil?
      return data if data.is_a?(String)

      JSON.generate(data)
    rescue JSON::GeneratorError => e
      raise CodecError, "JSON encode failed: #{e.message}"
    end

    def decode(data)
      return nil if data.nil? || data.empty?

      JSON.parse(data)
    rescue JSON::ParserError => e
      raise CodecError, "JSON decode failed: #{e.message}"
    end
  end

  # Binary payload codec (passthrough).
  class BinaryCodec
    include PayloadCodec

    def encode(data)
      unless data.is_a?(String)
        raise CodecError, "BinaryCodec expects String, got #{data.class}"
      end
      data.b
    end

    def decode(data)
      data.b
    end
  end

  # Null codec — encodes everything as empty string.
  class NullCodec
    include PayloadCodec

    def encode(_data)
      ''
    end

    def decode(_data)
      nil
    end
  end

  # Chain multiple codecs together.
  #
  # Encoding applies codecs left-to-right; decoding applies right-to-left.
  class CodecChain
    include PayloadCodec

    # @param codecs [Array<PayloadCodec>] Codecs to chain
    def initialize(*codecs)
      raise ArgumentError, 'CodecChain requires at least one codec' if codecs.empty?
      @codecs = codecs
    end

    def encode(data)
      result = data
      @codecs.each do |codec|
        result = codec.encode(result)
      end
      result
    end

    def decode(data)
      result = data
      @codecs.reverse.each do |codec|
        result = codec.decode(result)
      end
      result
    end
  end
end
